File: src/p7-reverse-integer.js
/*
 * @lc app=leetcode id=7 lang=javascript
 *
 * [7] Reverse Integer
 */

// @lc code=start
const LEFT_EDGE = -Math.pow(2, 31);
const RIGHT_EDGE = Math.pow(2, 31) - 1;

/**
 * @param {number} x
 * @return {number}
 */

const reverse = (x) => {
    let y = x < 0 ? -x : x;
    let result = 0;
    while (y > 0) {
        result = result * 10 + (y % 10);
        y = (y / 10) >> 0;
    }
    result = x < 0 ? -result : result;
    if (result < LEFT_EDGE || result > RIGHT_EDGE) {
        return 0;
    }

    return result;
};
// @lc code=end

[reverse(123), reverse(-123), reverse(120)].forEach((v) => console.log(v));

File: src/p7_reverse_integer.rs
/*
 * @lc app=leetcode id=7 lang=rust
 *
 * [7] Reverse Integer
 */
use super::Solution;

// @lc code=start
impl Solution {
    pub fn reverse(x: i32) -> i32 {
        let negative = x < 0;
        let mut y = if negative { -x } else { x };
        let mut result = 0i32;
        while y > 0 {
            if let Some(r) = result.checked_mul(10).and_then(|r| r.checked_add(y % 10)) {
                result = r;
                y /= 10;
            } else {
                return 0;
            }
        }

        if negative {
            -result
        } else {
            result
        }
    }
}
// @lc code=end

#[test]
fn test() {
    assert_eq!(Solution::reverse(123), 321);
    assert_eq!(Solution::reverse(-321), -123);
    assert_eq!(Solution::reverse(120), 21);
}
